deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java

```diff
@@ -107,16 +159,26 @@ public void applyUpdate(StepFunction function, INDArray params, INDArray updates
         // nullify given updates first
         updates.assign(0.0f);
 
+        int cnt = 0;
         while (!messages.get(index.get()).isEmpty()) {
             INDArray compressed = messages.get(index.get()).poll();
 
-            //log.info("Thread: {}; Compressed: {}", Thread.currentThread().getId(), Arrays.toString(compressed.data().asInt()));
-
-            // FIXME: pass gradients here, and reuse them
             INDArray decoded = Nd4j.getExecutioner().thresholdDecode(compressed, updates);
+            cnt++;
         }
 
-        function.step(params, updates);
+        if (externalSource != null)
+            while (!externalSource.isEmpty()){
+                INDArray compressed = externalSource.poll();
+
+                INDArray decoded = Nd4j.getExecutioner().thresholdDecode(compressed, updates);
+                cnt++;
+            }
+
+        // TODO: average updates probably?
```

Comment:
I'm totally, literally - totally - agree with you. Averaging is needed for performance reasons as well. I'm half way there.